src/se/terhol/test34/HouseImpl.java

```diff
@@ -0,0 +1,92 @@
+package se.terhol.test34;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Scanner;
+import java.util.Set;
+
+/**
+ * @author Tereza Holm
+ */
+public class HouseImpl implements House {
+    Map<Room, Set<Room>> myHouse = new HashMap<>();
+    Map<Person, Room> currentHouse = new HashMap<>();
+
+    public HouseImpl(String input) throws IOException {
+        InputStream is = Files.newInputStream(Paths.get(input));
+        this.loadPlan(is);
+    }
+
+    @Override
+    public void placePerson(Person person, Room room) {
+        if (person == null) {
+            throw new IllegalArgumentException();
+        }
+        currentHouse.put(person, room);
+    }
+
+    @Override
+    public Room whereIs(Person person) {
+        return currentHouse.get(person);
+    }
+
+    @Override
+    public void enterNextDoor(Person person, Room room) {
+        if (!currentHouse.containsKey(person) || !this.isNext(currentHouse.get(person), room)) {
+            throw new IllegalArgumentException();
+        }
+        this.placePerson(person, room);
+    }
+
+    @Override
+    public void savePlan(OutputStream os) {
+        try (BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(os))) {
+            for (Room room : myHouse.keySet()) {
+                for (Room roomNext : myHouse.get(room)) {
+                    writer.write(room.toString() + ":" + roomNext.toString());
+                    writer.newLine();
+                }
+            }
+        } catch (IOException e) {
+            System.err.println("IO error");
+        }
+    }
+
+    @Override
+    public void loadPlan(InputStream is) {
+        Scanner reader = new Scanner(is);
+
+        String line;
+
+        while (reader.hasNextLine()) {
+            line = reader.nextLine();
+            int spaceInt = line.indexOf(":");
+            Room room1 = new Room(line.substring(0, spaceInt));
+            Room room2 = new Room(line.substring(spaceInt + 1));
+
+            if (!myHouse.containsKey(room1)) {
```

Comment:
It is maybe a good idea to introduce some private method for 

```java
            if (!myHouse.containsKey(room1)) {
                myHouse.put(room1, new HashSet<>());
            }
            myHouse.get(room1).add(room2);
```

so you can avoid repeating the same code. It is also god to keep in mind that once a method contains more than 3 input parameters or more than five lines, it is good time to refactor it and split into more (private) method to achieve a better code readability.

Reply:
Done